cmd/neofs-node/container.go

```diff
@@ -79,11 +85,129 @@ func initContainerService(c *cfg) {
 		})
 	}
 
-	// TODO: report more often than once: https://github.com/nspcc-dev/neofs-node/issues/3543.
-	addNewEpochAsyncNotificationHandler(c, func(e event.Event) {
-		ev := e.(netmapEv.NewEpoch)
+	initSizeLoadReports(c)
+
+	cnrSrv := containerService.New(&c.key.PrivateKey, c.networkState, c.cli, (*containersInChain)(&c.basics), c.nCli)
+	addNewEpochAsyncNotificationHandler(c, func(event.Event) {
+		cnrSrv.ResetSessionTokenCheckCache()
+	})
+
+	for _, srv := range c.cfgGRPC.servers {
+		protocontainer.RegisterContainerServiceServer(srv, cnrSrv)
+	}
+}
+
+func initSizeLoadReports(c *cfg) {
+	l := c.log.With(zap.String("component", "containerReports"))
+
+	// hardcoded in neofs-contracts as of 957152c7da6eb0a20745cfa1c4513b3c3512db6e
+	const maxReportsPerEpoch = 3
+
+	nm, err := c.nCli.NetMap()
+	if err != nil {
+		fatalOnErr(fmt.Errorf("failed to get netmap to initialize container service: %w", err))
+	}
+	durSec, err := c.nCli.EpochDuration()
+	if err != nil {
+		fatalOnErr(fmt.Errorf("failed to get epoch duration to initialize container service: %w", err))
+	}
+	dur := time.Duration(durSec) * time.Second
+
+	var nmLen = len(nm.Nodes())
+	indexInNM := slices.IndexFunc(nm.Nodes(), func(node netmapsdk.NodeInfo) bool {
+		return bytes.Equal(node.PublicKey(), c.binPublicKey)
+	})
+	if indexInNM == -1 {
+		indexInNM = nmLen
+	}
+
+	var stepsInEpoch uint32
+	if nmLen != 0 {
+		// likely there is no need to report less often than once per second
```

Comment:
@roman-khimov, do you agree with it? another suggestion is block time as the first arg in `min`

Reply:
It's ok.